### ISSUE TITLE
[Extension Proposal] Add Graal Languages Extension to easily inject Python, JS, and WASM modules

### ISSUE THREAD

**Issue by timfel** [quarkusio/quarkus#43410]
### Description

Interested in this extension, please +1 via the emoji/reaction feature of GitHub (top right).

Since release 24.1.0, the Graal languages no longer need to be put on the Micronaut path, so they work with fat jars and on the classpath just fine. Micronaut has [added](https://guides.micronaut.io/latest/micronaut-graalpy.html) some support for GraalPy already that makes it easy to include Python modules.

It would be great to see something similar to this in Quarkus.

Some relevant links here are that GraalPy has some [utils](https://github.com/oracle/graalpython/blob/master/graalpython/org.graalvm.python.embedding.tools/src/org/graalvm/python/embedding/tools/vfs/VFSUtils.java) to manage putting Python packages into the right place inside Java resources. These are used by the [JBang](https://github.com/oracle/graalpython/blob/0bbb8bce4e680243751e5432d099132f0eb117b0/graalpython/org.graalvm.python.embedding.tools/src/org/graalvm/python/embedding/tools/jbang/JBangIntegration.java#L46), [Gradle](https://github.com/oracle/graalpython/blob/0bbb8bce4e680243751e5432d099132f0eb117b0/graalpython/org.graalvm.python.gradle.plugin/src/main/java/org/graalvm/python/GraalPyGradlePlugin.java#L61) and [Maven](https://github.com/oracle/graalpython/blob/0bbb8bce4e680243751e5432d099132f0eb117b0/graalpython/graalpy-maven-plugin/src/main/java/org/graalvm/python/maven/plugin/ManageResourcesMojo.java#L62) plugins already, and are considered public API. The Micronaut integration is [here](https://github.com/micronaut-projects/micronaut-graal-languages/tree/1.0.x/graalpy/src/main/java/io/micronaut/graal/graalpy) and has all the classes that implement the injection.

### Repository name

quarkus-graal-languages

### Short description

Use Graal Languages in Quarkus

### Repository Homepage URL

https://quarkiverse.github.io/quarkiverse-docs/<REPOSITORY_NAME>/dev/

### Repository Topics

- quarkus-extension
- 
...

### Team Members

- 

### Additional context

We are happy to help from the GraalPy team, but we do not have the resources to lead the implementation in this.

**Comment by maxandersen** [comment]
@jponge @geoand fyi

**Comment by jponge** [comment]
@timfel Does GraalVM support host functions? I've started to play a bit and it's quite easy to run some guest functions from WASM, but I haven't figured out how to expose host functions to WASM. Thanks!

**Comment by jponge** [comment]
Thanks I'll have a look!

**Comment by LucienBrule** [comment]
Greetings all,

I’ve put together a working prototype ([quarkus-graalpy-demo](https://github.com/LucienBrule/quarkus-graalpy-demo)) that embeds GraalPy in dev/JVM/native modes and unblocks native builds via the fix in [#47666](https://github.com/quarkusio/quarkus/issues/47666).

I’d like to take the next step and spearhead a **Quarkiverse `quarkus-graal-languages` extension** offering:

- CDI-managed isolates for Python and JavaScript (initial focus)  
- DevMode hot-reload plus a DevUI REPL  
- Native-image configuration baked in  
- Helper APIs to sandbox Model Context Protocol servers and other polyglot tooling

If that direction aligns with Quarkus’ roadmap, could we spin up a Quarkiverse repo for the extension? I’m ready to maintain and iterate in the open—and welcome collaborators.

Related proposal: quarkusio/quarkus#43410 (this one)

/cc @maxandersen @geoand @jponge @timfel

**Comment by maxandersen** [comment]
This looks quite interesting!

**Comment by maxandersen** [comment]
looking at the example repo it seem to be missing `python-resources/` ?

**Comment by geoand** [comment]
Very interesting indeed!

**Comment by fniephaus** [comment]
Very cool!

Would be nice to have something like this as a [GraalPy demo](https://github.com/graalvm/graal-languages-demos/tree/main/graalpy). If you prefer a separate repository, we could also link to it directly from the GraalPy website.

**Comment by jponge** [comment]
@LucienBrule sounds very cool indeed!

Quick question: what are the multi-threading assumptions around the Graal `Context` abstraction? I saw that you have an application-scoped component, but my understanding is that a `Context` might not be thread-safe, and at the very least we might not want to always re-use a `Context` for different sessions.

**Comment by maxandersen** [comment]
https://medium.com/graalvm/whats-new-in-graal-languages-24-2-328471fc4137

Says one context per thread. 

@fniephaus or @timfel can probably best comment on best practice here. 

And if there are some virtual thread considerations too...

**Comment by timfel** [comment]
> Quick question: what are the multi-threading assumptions around the Graal Context abstraction?

There is no easy answer to this, since it depends on the semantics of the underlying languages. You can use contexts from multiple threads, the APIs are perfectly thread-safe. It's the language runtimes that impose constraints.

For example, JavaScript as a language has no concept of true multi-threading, so accessing the same Graal.js context from multiple Java threads *at the same time* will fail. Python, on the other hand, does allow multi-threaded access, but the semantics and compatibility constraints dictate a global interpreter lock. Thus, if two Java threads want to evaluate Python code, they _might_ get blocked on the Python global interpreter lock.

You can just use multiple contexts and [share the Engine](https://www.graalvm.org/latest/reference-manual/embed-languages/#code-caching-across-multiple-contexts), this way each context executes truly in parallel and their global state is isolated from on another. *But* now you may run into problems if your GraalPy code uses native extensions, which rely on loading real native libraries. The mainstream operating systems do not allow us to simply load native libraries multiple times into the same process and keep them isolated from each other. So for GraalPy, either you stick with a single context per application in this case _or_ you use the [`python.IsolateNativeModules`](https://medium.com/graalvm/whats-new-in-graal-languages-24-2-328471fc4137) option we recently introduced.

**Comment by fniephaus** [comment]
I'd recommend going for the single context approach by default, which is easy to explain to users. In a next step, you could allow users to configure their setup, e.g. pool of contexts (for throughput), or one context per request (for isolation).

**Comment by geoand** [comment]
+1 for moving this ahead ASAP

**Comment by timfel** [comment]
> And if there are some virtual thread considerations too...

Missed this one. There are - you cannot use virtual threads and native code with GraalPy at the same time. If you are lucky your native thread will simply get pinned to an OS thread, because of the native call. But if you're unlucky, the native library that is (maybe transitively) used by the Python extension uses a native thread local somewhere. And then you end up leaking thread local data between virtual threads, which can lead to hard-to-debug errors and crashes.

My recommendation would be to expose the `allowNativeAccess` context option in your GraalPyConfig. This way, when people opt in you know you cannot use virtual threads, but you can still use them if people do not opt in. The users will notice very quickly if they are importing native extensions :)

**Comment by maxandersen** [comment]
The limits for native libs applies equally to java so doesn't sound like what would need to be treated differently. 

I would have expected use one environment but different context per execution context (by thread or some other key) to avoid global serialization ?

**Comment by andreaTP** [comment]
I missed the chance to join this thread earlier. I'm currently working on something closely related, though with very different trade-offs, based on WASM.

Specifically, I'm working on a JavaScript engine for Java: [QuickJs4J](https://github.com/roastedroot/quickjs4j)

QuickJS is extremely lightweight. It’s not thread-safe, but we can spin up multiple instances (we'll probably need to run some benchmarks to find the optimal configuration). Notably, it doesn't depend on any global or native setup.

Before diving into technical details, I’d love to briefly touch on a couple of Developer Experience design points.

Providing just a single eval function (à la JSR 223) is a very low-level API as it immediately exposes users to the complexities of cross-language interaction.

If one of the goals here is to enable users to write "plugins" we should make it easy for developers to expose Java APIs to the scripting context (e.g., JS or Python). In [QuickJs4J](https://github.com/roastedroot/quickjs4j), I introduced the concept of `Builtins` for this purpose.

The second point is that developers rarely want to drop a raw JS/Python String into their code, as seen in many demos. Instead, they'll likely use build and packaging tools and want to reuse existing language-native tooling. When invoking a JS/Python function from Java, developers expect a tighter integration that automatically maps native Java types. This is something I’ve tentatively modeled as `Invokables` in [QuickJs4J](https://github.com/roastedroot/quickjs4j).

cc. @EricWittmann  @chirino

**Comment by EricWittmann** [comment]
+1 @andreaTP 

We're still iterating, but you can see progress here:

https://github.com/EricWittmann/quarkus-script-playground

I've been exploring a streamlined developer experience where you can just define an interface and annotate it as a javascript SPI:

https://github.com/EricWittmann/quarkus-script-playground/blob/main/quarkus-test-application/src/main/java/io/apicurio/calculator/Calculator.java

You can then just implement the interface in a javascript module:

https://github.com/EricWittmann/quarkus-script-playground/blob/main/quarkus-test-application/src/main/ts/Calculator.ts

And then just inject it:

https://github.com/EricWittmann/quarkus-script-playground/blob/main/quarkus-test-application/src/main/java/io/apicurio/app/SingleCalculatorService.java

**Comment by timfel** [comment]
> The limits for native libs applies equally to java so doesn't sound like what would need to be treated differently.

👍 

> I would have expected use one environment but different context per execution context (by thread or some other key) to avoid global serialization ?

If you somehow expose whether or not native libraries are allowed in the Python context, then you can make the decision in the framework: multiple [contexts](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/Context.html) with a shared [engine](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/Engine.html) running on virtual threads (maybe a context pool and one context per request?). This should give you the highest throughput because there is no global interpreter lock between isolated contexts. Only when native libraries are allowed I would go conservative: single global context + platform threads
